### PR TITLE
Implemented marker-based pagination in `box_ls`

### PR DESCRIPTION
**Pull request by awong234**
# Problem: [Issue #74]

Offset-based pagination has an offset limit of 300,000 which is not stated in the Box API documentation, but will be added following [conversation with the Box team](https://community.box.com/t5/custom/page/page-id/BoxViewTicketDetail?ticket_id=1685576). The original function for `box_pagination` has been renamed as `box_paginate_offset` to preserve original functionality.

# Solution

The team recommended the use of marker-based pagination. Marker-based pagination is, according to the Box API docs, ["the preferred method and is most performant"](https://developer.box.com/reference#get-a-folders-items), and so this change makes marker-based pagination the default method for GET-ting folder items. 

# Tests

My tests succeed in comparing and syncing my local set of files to the remote set of my box folder with quantity of files >330,000 where offset-based pagination returns an HTTP 400 error. Microbenchmark of the tests show comparable time to query 100,000 file names

```
Unit: seconds
   expr      min       lq     mean   median       uq      max neval
 OFFSET 37.52012 40.04692 41.27804 40.93159 41.68023 50.97874   100
 MARKER 40.87141 42.31870 44.24652 42.98368 43.90257 59.99026   100
```

### CODE REVIEW

**Comment by ijlyttle** [comment]
Hi Alec,

I appreciate this PR. Unfortunately it may be a few days before I will have time to look at it properly.

Ian

**Comment by awong234** [comment]
Certainly, take your time! I will let you know if I encounter any issues within my fork.

File: R/boxr_misc.R [hunk 1]
```diff
@@ -49,15 +61,64 @@ box_ls <- function(dir_id = box_getwd(), limit = 100, max = Inf, fields = NULL)
...
+  
+  marker = character(0)
+  
+  n_so_far = 0
+  
+  out = list()
+  
+  url$query$usemarker = T
+  
+  while(!is.null(marker)){
```
Comment (ijlyttle) [inline R/boxr_misc.R]: I would use a logical variable here, `next_page`, like `box_paginate_offset()`.

**Comment by awong234** [comment]
Hi Ian,

I agree with dropping support for offset-based pagination; there was no benefit that I could see other than preserving legacy function. I will get these changed soon. Thanks!

**Comment by awong234** [comment]
Hi Ian, 

Wasn't sure if you were notified of this - I made the changes requested. Thanks for reviewing the code!

**Comment by ijlyttle** [comment]
Hi Alec,

I hope to get to this in the next couple of days, thanks!

**Comment by awong234** [comment]
Ok great thanks! I've had a few of my users use my fork without errors so
far, I will let you know if we encounter anything.

File: R/boxr_misc.R [hunk 2]
```diff
@@ -1,36 +1,39 @@
...
+#'
+#' @seealso [box_fetch()] and [box_push()] for synchronizing the contents of
+#'   local and remote directories. [list.files()] for examining the contents of
+#'   local directories.
+#'
 #' @export
 box_ls <- function(dir_id = box_getwd(), limit = 100, max = Inf, fields = NULL) {
   
-  # maybe some logic here to check that limit <= 1000
+  if(limit > 1000){
```
Comment (ijlyttle) [inline R/boxr_misc.R]: Logic looks good here - could you format to match the spacing and indentation of the code at L41, and use `<-` for assignment?

Could you do this for the other logic blocks (`if`, `while`) and use `<-` for assignment, generally?

**Comment by awong234** [comment]
Thanks Ian! I will get to these soon; I'm just returning from travel so I can address the last bits now.

**Comment by awong234** [comment]
Hi Ian,

I made those changes that you requested. 

I also took the liberty to make the 404 error a little more explicit (see [12834e4](https://github.com/brendan-r/boxr/pull/79/commits/12834e4e9ddda3394bcd2706ec8ad264ecbdbcc3); I had a user try to connect to a shared folder that they were not a member of, and I think the extra information could have made me see the problem sooner. 

Thanks for the review! 

Alec

**Pull request merged**